README.md
# @cionzo/js-changelog 1.3.1 

Semantic versioning and changelog generation utils

## Installation

This is a [Node.js](https://nodejs.org/) module available through the [github repository](https://www.github.com/).

It can be installed using the 
[`npm`](https://docs.npmjs.com/getting-started/installing-npm-packages-locally)
or 
[`yarn`](https://yarnpkg.com/en/)
command line tools.

```sh
npm install git+https://github.com/cionz0/js-changelog --save-dev
```

## Usage

### Configuration
Add the target `js-changelog` to the `scripts` section of your `package.json` file.

```JSON
{
    // ...
    "scripts": {
        // ...
        "js-changelog": "npx @cionzo/js-changelog"
    }
}
```

Remember that, in order to pass arguments to a node-invoked target, you should use `--`.
```sh
node js-changelog -- --help
```
### Example 

In order to correctly use this tool, your changes must be already committed and pushed.
Your local and remote repositories are hence assumed to refer to the same commit.

At this point, you can run something like the following command.
```sh
node js-changelog -- --minor
```

## Tests

```sh
npm install
npm test
```

## Dependencies

- [argparse](https://ghub.io/argparse): CLI arguments parser. Native port of python&#39;s argparse.
- [auto-changelog](https://ghub.io/auto-changelog): Command line tool for generating a changelog from git tags and commit history
- [prompt-sync](https://ghub.io/prompt-sync): a synchronous prompt for node.js


## License

GPL-3.0


CHANGELOG.md
# Unreleased
[8 September 2023]

### Other

* 2023-09-08T19:30:30.000Z [cionzo] - other: updated readme ([`a3710bb`](https://github.com/cionz0/js-changelog/commit/a3710bbfa417b8357f1f75778334b671290c168b))
* 2023-09-08T19:31:19.000Z [cionzo] - other: updated readme ([`1ee7fe0`](https://github.com/cionz0/js-changelog/commit/1ee7fe0864e18a640131203445593cc459f72b25))
* 2023-09-08T19:31:29.000Z [cionzo] - version bump: 1.3.1 ([`be79ae5`](https://github.com/cionz0/js-changelog/commit/be79ae5bf96f9fd5c7e5b60d629ff332210987e9))
# 1.3.0
[8 September 2023]

### Fixes

* 2023-09-08T19:23:20.000Z [cionzo] - fix: solved git issues ([`09db784`](https://github.com/cionz0/js-changelog/commit/09db784fa3704081f9d063638d4f3909d7d47d52))
* 2023-09-08T07:06:34.000Z [cionzo] - fix: set verbose mode when cli options ask for current release ([`d736d0a`](https://github.com/cionz0/js-changelog/commit/d736d0a23b65d05d7f9b138ef8c87a796d7f55c1))

### Other

* 2023-09-08T19:26:04.000Z [cionzo] - version bump: 1.3.0 ([`76b3c79`](https://github.com/cionz0/js-changelog/commit/76b3c79d298f2175efe815b6e5f91fd7c6236ea9))
# 1.2.2
[8 September 2023]

### Fixes

* 2023-09-08T07:06:34.000Z [cionzo] - fix: set verbose mode when cli options ask for current release ([`895e853`](https://github.com/cionz0/js-changelog/commit/895e853e4f8402283335c9bb2478a565073f7239))

# 1.2.1
[8 September 2023]

### Fixes

* 2023-09-08T06:54:56.000Z [cionzo] - fix: set action object for currentRelease ([`93f98de`](https://github.com/cionz0/js-changelog/commit/93f98de9459f04777d432883bb3199a53b9a27f1))
* 2023-09-08T06:54:56.000Z [cionzo] - fix: set action object for currentRelease ([`14c5d70`](https://github.com/cionz0/js-changelog/commit/14c5d70c6d4c2c480440e6fee4cfb3b9f6677855))

### Other

* 2023-09-07T19:22:30.000Z [cionzo] - feat: just to test ([`c956560`](https://github.com/cionz0/js-changelog/commit/c9565605fcc9530e6af1dd0d7cf7ed86b9d3cc39))
# 1.2.0
[7 September 2023]

### Other

* 2023-09-07T19:17:44.000Z [cionzo] - feat: working ([`89fe5b3`](https://github.com/cionz0/js-changelog/commit/89fe5b30e22f93403b02827f7ffb22dc65bdc9fe))
* 2023-09-07T19:19:56.000Z [cionzo] - feat: just to test ([`f5f822b`](https://github.com/cionz0/js-changelog/commit/f5f822ba6473d66641adcf40c4f4afb11655c8ed))
# 1.1.0
[7 September 2023]

### Other

* 2023-09-07T19:14:20.000Z [cionzo] - feat: working ([`4bf9d73`](https://github.com/cionz0/js-changelog/commit/4bf9d73f7483adb377eccf510ab3adbdff92ad7d))
* 2023-09-07T17:33:05.000Z [cionzo] - other: cleanings ([`1774976`](https://github.com/cionz0/js-changelog/commit/1774976aeabb5a7986666d37e12dc8c97b20a91f))
* 2023-09-07T19:17:44.000Z [cionzo] - feat: working ([`11b94f6`](https://github.com/cionz0/js-changelog/commit/11b94f686a9b5bde559984ad5d1f2db744efa9c2))
# 1.0.0
[7 September 2023]

### Other

* 2022-05-23T09:38:54.000Z [cionz0] - Initial commit ([`5963235`](https://github.com/cionz0/js-changelog/commit/59632357bf9b0db8d8685e1b80f2ac55a7da5fea))
* 2023-09-07T17:33:05.000Z [cionzo] - other: cleanings ([`a3111e9`](https://github.com/cionz0/js-changelog/commit/a3111e90effb63e35c2cffa893f1ef0b1962f59b))
* 2023-09-07T11:56:11.000Z [cionzo] - initial commit ([`d1da803`](https://github.com/cionz0/js-changelog/commit/d1da803963e900644fae548b63a8f7a63b699b33))


eslint.js
module.exports = {
    // extensions: ['.js'],
    parserOptions: {
        ecmaVersion: 2020,
    },
    env: {es6: true, node: true},
    extends: ["eslint:recommended"],
    plugins: ["jsdoc"],
    // settings: {
    //     "jsdoc": {
    //         "tagNamePreference": {
    //             "return": {"message": "please use returns instead"},
    //         },
    //     },
    // },
    rules: {
        // 'no-extra-semi': 0,
        "no-console": 0,
        "indent": ["error", 4, {"SwitchCase": 1}],
        //"brace-style": ["error", "stroustrup", { "allowSingleLine": true }]
        "quotes": ["error", "double"],
        "max-len": [1, {code: 900}],
        "camelcase": [0],
        // "valid-jsdoc": [1],
        "require-jsdoc": [1],
        "new-cap": [1],
        "no-invalid-this": [1],
        "no-undef": [1],
        "prefer-const": [1],
        "guard-for-in": [1],
        "jsdoc/check-access": 1, // Recommended
        "jsdoc/check-alignment": 1, // Recommended
        // "jsdoc/check-examples": 1,
        "jsdoc/check-indentation": [1, {"excludeTags": ["return", "returns", "openapi", "example"]}],
        "jsdoc/check-line-alignment": 1,
        "jsdoc/check-param-names": 1, // Recommended
        "jsdoc/check-property-names": 1, // Recommended
        "jsdoc/check-syntax": 1,
        "jsdoc/check-tag-names": ["warn", {"definedTags": ["project", "return", "openapi", "createdOn"]}], // Recommended
        "jsdoc/check-types": 1, // Recommended
        "jsdoc/check-values": 1, // Recommended
        "jsdoc/empty-tags": 1, // Recommended
        "jsdoc/implements-on-classes": 1, // Recommended
        "jsdoc/match-description": 1,
        "jsdoc/multiline-blocks": 1, // Recommended
        // "jsdoc/newline-after-description": 1, // Recommended
        "jsdoc/no-bad-blocks": 1,
        "jsdoc/no-defaults": 1,
        // "jsdoc/no-missing-syntax": 1,
        "jsdoc/no-multi-asterisks": 1, // Recommended
        // "jsdoc/no-restricted-syntax": 1,
        // "jsdoc/no-types": 1,
        "jsdoc/no-undefined-types": 1, // Recommended
        "jsdoc/require-asterisk-prefix": 1,
        "jsdoc/require-description": 1,
        "jsdoc/require-description-complete-sentence": 1,
        // "jsdoc/require-example": 1,
        // "jsdoc/require-file-overview": 1,
        "jsdoc/require-hyphen-before-param-description": 1,
        "jsdoc/require-jsdoc": 1, // Recommended
        "jsdoc/require-param": 1, // Recommended
        "jsdoc/require-param-description": 1, // Recommended
        "jsdoc/require-param-name": 1, // Recommended
        "jsdoc/require-param-type": 1, // Recommended
        "jsdoc/require-property": 1, // Recommended
        "jsdoc/require-property-description": 1, // Recommended
        "jsdoc/require-property-name": 1, // Recommended
        "jsdoc/require-property-type": 1, // Recommended
        "jsdoc/require-returns": 1, // Recommended
        "jsdoc/require-returns-check": 1, // Recommended
        "jsdoc/require-returns-description": 1, // Recommended
        "jsdoc/require-returns-type": 1, // Recommended
        "jsdoc/require-throws": 1,
        "jsdoc/require-yields": 1, // Recommended
        "jsdoc/require-yields-check": 1, // Recommended
        "jsdoc/sort-tags": 1,
        "jsdoc/tag-lines": 1, // Recommended
        "jsdoc/valid-types": 1, // Recommended
        "valid-jsdoc": [1, {
            "prefer": {
                "return": "returns",
            },
        }],

    },

    ignorePatterns: ["docs/jsdoc/*"],
}


src/index.js
#!/usr/bin/env node
/**
 * @module js-changelog
 * @description A command-line tool for managing Git tags and versioning.
 * @createdOn 2023-09-07
 * @project js-changelog
 * @author cionzo
 */

"use strict";

// Import required modules
const CLI_COMMANDS = require("./cli_commands");
const {CommandLineArgs} = require("./type_definitions")
const support = require("./support")
const parsing = require("./parsing")
// const constants = require("./constants");



//
//
// /**
//  * Perform a major release, including tagging and updating the changelog.
//  * @param {CommandLineArgs} args - Command line arguments.
//  * @returns {Promise<void>} A Promise that resolves when the major release process is completed successfully.
//  * @throws {Error} If an error occurs during the release process.
//  */
// async function majorRelease(args) {
//     await createRelease(RELEASE_TYPES.MAJOR, args);
// }
//
// /**
//  * Perform a minor release, including tagging and updating the changelog.
//  * @param {CommandLineArgs} args - Command line arguments.
//  * @returns {Promise<void>} A Promise that resolves when the minor release process is completed successfully.
//  * @throws {Error} If an error occurs during the release process.
//  */
// async function minorRelease(args) {
//     await createRelease(RELEASE_TYPES.MINOR, args);
// }
//
// /**
//  * Perform a patch release, including tagging and updating the changelog.
//  * @param {CommandLineArgs} args - Command line arguments.
//  * @returns {Promise<void>} A Promise that resolves when the patch release process is completed successfully.
//  * @throws {Error} If an error occurs during the release process.
//  */
// async function patchRelease(args) {
//     await createRelease(RELEASE_TYPES.PATCH, args);
// }








/**
 * Update the changelog for a given version.
 * @param {string} versionTagStr - The version tag for which the changelog is updated.
 * @param {CommandLineArgs} args - Command line arguments.
 * @returns {Promise<void>} A promise that resolves when the changelog is updated.
 */
async function updateChangelog(args, versionTagStr) {
    if (!versionTagStr) {
        versionTagStr = await support.currentRelease(args);
    }
    versionTagStr = versionTagStr.trim()
    await CLI_COMMANDS.execute(`auto-changelog --unreleased --template ${args.changelog_template} --output ${args.changelog_output}`);
    console.log(`Changelog for version ${versionTagStr} created.`);
}

/**
 * Create a new release, including tagging and updating the changelog.
 * @param {CommandLineArgs} args - Command line arguments.
 * @returns {Promise<void>} A Promise that resolves when the release creation process is completed successfully.
 * @throws {Error} If an error occurs during the release process.
 */
async function createRelease(args) {
    try {
        const newTag = await support.computeNewReleaseTag(args);
        console.log(`new release tag is ${newTag}`)
        await updatePackageJsonVersion(newTag, args);
        await support.tagCurrentCommit(newTag);
        await updateChangelog(args, newTag);
    } catch (error) {
        throw new Error(`Failed creating ${args.action.release} release: ${error.message}`);
    }
}



/**
 * Update the version in package.json.
 * @param {string} newRel - The new release version.
 * @param {CommandLineArgs} args - Command line arguments.
 * @returns {Promise<void>} A promise that resolves when the package.json is updated.
 */
async function updatePackageJsonVersion(newRel, args) {
    const fs = require("fs");
    const PACKAGE_JSON_DATA = JSON.parse(fs.readFileSync(args.package_json).toString());
    PACKAGE_JSON_DATA.version = newRel.trim();
    fs.writeFileSync(args.package_json, JSON.stringify(PACKAGE_JSON_DATA, undefined, 4));

    await CLI_COMMANDS.execute(`git pull`);
    await CLI_COMMANDS.execute(`git add ${args.package_json}`);
    await CLI_COMMANDS.execute("git commit --amend --no-edit");
    await CLI_COMMANDS.execute("git push");
}



module.exports = {

    updateChangelog,
    createRelease,
    currentRelease: support.currentRelease
};

// Handle command-line execution
if (require.main === module) {

    const parser = parsing.getParser()
    const args = parser.parse_args();

    parsing.processArgs(args)

    // Call the appropriate function based on the selected action

    args.action ? module.exports[`${args.action.name}`](args) : parser.print_help();
}


src/support.js
/**
 * This module ...
 *
 * <!-- do not type below this line -->
 * Created on: 07/09/23.
 * @module src/support
 * @project js-changelog
 * @author cionzo <cionzo@filotrack.com>
 */
"use strict"

const CLI_COMMANDS = require("./cli_commands");
const constants = require("./constants")
const {SemVerInfo, CommandLineArgs} = require("./type_definitions")



/**
 * Check if the 'git' command is available in the system.
 * @returns {Promise<boolean>} A promise that resolves to true if the 'git' command is available; otherwise, false.
 * @throws {Error} If the 'git' command is not available.
 * @private
 */
async function isGitCommandAvailable() {
    try {
        await CLI_COMMANDS.execute("git --version");
        return true;
    } catch (error) {
        throw new Error("Git command is unavailable. Please install Git or add it to your PATH.");
    }
}

/**
 * Get the associated tag for a given commit SHA.
 * @param {string} commitSHA - The commit SHA to find the associated tag.
 * @returns {Promise<string|null>} A promise that resolves to the associated tag or null if no tag is found.
 * @private
 */
async function getAssociatedTag(commitSHA) {
    try {
        return await CLI_COMMANDS.execute(`git describe --tags ${commitSHA}`);
    } catch (error) {
        return null; // No associated tag found
    }
}
/**
 * Get the SHA of the last tagged commit.
 * @returns {Promise<string|null>} A promise that resolves to the SHA of the last tagged commit or null if no tags exist.
 * @private
 */
async function getLastTaggedCommitSHA() {
    try {
        return await CLI_COMMANDS.execute("git rev-list --tags --max-count=1");
    } catch (error) {
        return null; // No tags exist
    }
}


/**
 * Parse a version tag into its components.
 * @param {string} versionTag - The version tag to parse.
 * @returns {object} An object with 'description', 'major', 'minor', and 'patch' properties as numbers.
 * @throws {Error} If the version tag cannot be parsed.
 * @private
 */
function parseVersionTag(versionTag) {
    const match = constants.SEMVER_TAG_PATTERN.exec(versionTag);

    if (!match || !match.groups) {
        throw new Error(`Unable to parse version tag '${versionTag}'. It does not match the expected pattern.`);
    }

    const {description, major, minor, patch} = match.groups;

    return {description, major: Number(major), minor: Number(minor), patch: Number(patch)};
}

/**
 * Tag the current commit with a given tag.
 * @param {string} tag - The tag to apply to the current commit.
 * @returns {Promise<void>} A promise that resolves when the commit is tagged successfully.
 * @throws {Error} If the provided tag does not match the SEMVER format.
 * @private
 */
async function tagCurrentCommit(tag) {
    if (!tag.match(constants.SEMVER_TAG_PATTERN)) {
        throw new Error(`The provided tag (${tag}) does not conform to SEMVER format (${constants.SEMVER_TAG_PATTERN}).`);
    }
    await CLI_COMMANDS.execute(`git tag ${tag}`);
}



/**
 * Get the current release version or its components.
 * @param {CommandLineArgs} args - Command line arguments.
 * @param {boolean} [parsed] - If true, parse the version tag and return its components.
 * @returns {Promise<string|SemVerInfo>} A promise that resolves to the current release version as a string or an object
 * with 'description', 'major', 'minor', and 'patch' properties if 'parsed' is true.
 * @throws {Error} If the 'git' command is not available or if 'parsed' is true and the version tag cannot be parsed.
 */
async function currentRelease(args,parsed = false) {
    if (!await isGitCommandAvailable()) {
        throw new Error("Git command is unavailable. Please install Git or add it to your PATH.");
    }

    const lastTaggedCommitSHA = await getLastTaggedCommitSHA();
    const currentReleaseStr = lastTaggedCommitSHA ? await getAssociatedTag(lastTaggedCommitSHA) : constants.SEMVER_ZERO;

    if (parsed) {
        return parseVersionTag(currentReleaseStr);
    }

    if (args.verbose) {
        console.log(`Current release is ${currentReleaseStr}`);
    }
    return currentReleaseStr;
}

/**
 * Compute a new release tag based on the current version and the specified release type.
 * @param {CommandLineArgs} args - Command line arguments.
 * @returns {Promise<string>} A promise that resolves to the new release tag.
 * @throws {Error} If the current release cannot be parsed or if the release type is invalid.
 */
async function computeNewReleaseTag(args) {
    let current = "\"undefined\"";
    try {
        current = await currentRelease(args,true);
        let newVersion;

        switch (args.action.release) {
            case constants.RELEASE_TYPES.MAJOR:
                newVersion = `${current.description}${current.major + 1}.0.0`;
                break;
            case constants.RELEASE_TYPES.MINOR:
                newVersion = `${current.description}${current.major}.${current.minor + 1}.0`;
                break;
            case constants.RELEASE_TYPES.PATCH:
                newVersion = `${current.description}${current.major}.${current.minor}.${current.patch + 1}`;
                break;
            default:
                throw new Error(`Invalid release type: "${args.action.release}". It must be one of ${Object.keys(constants.RELEASE_TYPES)}.`);
        }

        return newVersion;
    } catch (error) {
        throw new Error(`Unable to compute a new release tag starting from ${current}.\n Args: ${args}.\nError: ${error.message}`);
    }
}

module.exports = {
    isGitCommandAvailable,
    getAssociatedTag,
    getLastTaggedCommitSHA,
    parseVersionTag,
    tagCurrentCommit,
    currentRelease,
    computeNewReleaseTag

}

src/constants.js
/**
 * This module ...
 *
 * <!-- do not type below this line -->
 * Created on: 07/09/23.
 * @module src/defaults
 * @project js-changelog
 * @author cionzo <cionzo@filotrack.com>
 */
"use strict"

/**
 * An object representing release types for version incrementation.
 * @constant {object}
 * @property {string} MAJOR - Represents a major version increment.
 * @property {string} MINOR - Represents a minor version increment.
 * @property {string} PATCH - Represents a patch version increment.
 */
const RELEASE_TYPES = {
    MAJOR: "major",
    MINOR: "minor",
    PATCH: "patch"
};


/**
 * Regular expression pattern for parsing SEMVER tags.
 * @type {RegExp}
 */
const SEMVER_TAG_PATTERN = /(?<description>[a-zA-Z]*)(?<major>\d+)\.(?<minor>\d+).(?<patch>\d+)\s*/g;

/**
 * Default SEMVER version when no tags are available.
 * @type {string}
 */
const SEMVER_ZERO = "0.0.0";

module.exports = {
    CHANGELOG_PATH: "./CHANGELOG.md",
    CHANGELOG_TEMPLATE_PATH: "./node_modules/@cionzo/js-changelog/configs/changelog-template.hbs",
    PACKAGE_JSON_PATH: "./package.json",
    RELEASE_TYPES,
    SEMVER_ZERO,
    SEMVER_TAG_PATTERN
}
 

src/cli_commands.js
/**
 * Module for executing command-line commands asynchronously.
 * @module src/cli_commands
 * @project js-changelog
 * @createdOn 07/09/23
 * @author cionzo <cionzoh@gmail.com>
 */

"use strict";

const { promisify } = require("util");
const exec = promisify(require("child_process").exec);

/**
 * Execute a command and return its standard output.
 * @param {string} command - The command to execute.
 * @returns {Promise<string>} A promise that resolves to the standard output of the executed command.
 */
async function execute(command) {
    try {
        const execution = await exec(command);
        return execution.stdout;
    } catch (error) {
        // Handle any execution errors here
        throw new Error(`Failed to execute command: ${command}\nError: ${error.message}`);
    }
}

module.exports = { execute };


src/type_definitions.js
/**
 * @module src/type_definitions
 * @description Contains custom type definitions for js-changelog.
 * @createdOn 2023-09-07
 * @project js-changelog
 * @author cionzo
 */

"use strict";

/**
 * @typedef {object} SemVerInfo
 * @description Represents semantic versioning information.
 * @property {string} description - A short description or prefix for the version.
 * @property {number} major - The major version number.
 * @property {number} minor - The minor version number.
 * @property {number} patch - The patch version number.
 */

/**
 * @typedef {object} CommandLineArgs
 * @description Represents command-line arguments for js-changelog.
 * @property {string} [action] - The action to perform (optional).
 * @property {string} changelog_output - The path to the changelog file.
 * @property {string} changelog_template - The path to the changelog template file.
 * @property {string} package_json - The path to the package.json file.
 * @property {boolean} verbose - Indicates whether to enable verbose output.
 */

module.exports = {
    SemVerInfo: null,
    CommandLineArgs: null
};


configs/release_utils.js
#!/usr/bin/env node
/**
 * This module is a helper for managing releases
 *
 * <!-- do not type below this line -->
 * Created on: 24/11/21
 * @module configs/release_utils
 * @author cionzo <cionzo@filotrack.com>
 */
"use strict"

module.exports = {
  major_release,
  minor_release,
  fix_release,
  current_release,
  _update_changelog,
}

const {ArgumentParser} = require('argparse')
const {promisify} = require('util')
const exec = promisify(require('child_process').exec)
const path = require('path')

let parser = new ArgumentParser({
  description: 'Manages releases for this project.',
  add_help: true,
})

parser.add_argument('PROJECT_NAME', {help: 'The project name'})
//
parser.add_argument('-v', '--verbose', {action: 'store_true'})
parser.add_argument('-c', '--current', {
  dest: 'action',
  action: 'store_const',
  const: 'current_release',
  help: 'returns the current release number',
})
parser.add_argument('-M', '--major', {
  dest: 'action',
  action: 'store_const',
  const: 'major_release',
  help: 'creates a new major release',
})
parser.add_argument('-m', '--minor', {
  dest: 'action',
  action: 'store_const',
  const: 'minor_release',
  help: 'creates a new minor release',
})
parser.add_argument('-f', '--fix', {
  dest: 'action',
  action: 'store_const',
  const: 'fix_release',
  help: 'creates a new fix release',
})

parser.add_argument('-cl', '--changelog', {
  dest: 'action',
  action: 'store_const',
  const: '_update_changelog',
  help: 'updates the changelog file',
})

const args = parser.parse_args()

function _print_verbose(_string) {
  if (args.verbose) console.log(_string)
}

async function _run_shell_command(command) {
  const execution = await exec(command)
  return execution.stdout
}

async function current_release() {
  const command = 'git describe --tags $(git rev-list --tags --max-count=1)' //'git describe --tags --abbrev=0'
  const output = await _run_shell_command(command)

  // const delme = re.compile(r'''(?P<description>[a-zA-Z]*)(?P<major>[0-9]+).(?P<minor>[0-9]+).(?P<fix>[0-9]+)''',re.VERBOSE)
  const pattern = /([a-zA-Z]*)(\d+)\.(\d+).(\d+)\s*/g
  let match = pattern.exec(output)
  let description = '',
    major = '',
    minor = '',
    fix = ''
  console.log(output)
  description = match[1]
  major = parseInt(match[2])
  minor = parseInt(match[3])
  fix = parseInt(match[4])

  _print_verbose(`Current release is: ${description}${major}.${minor}.${fix}`)
  return {description, major, minor, fix}
}

async function major_release() {
  let {description, major, minor, fix} = await current_release()
  major += 1
  minor = 0
  fix = 0
  let release_tag = `${description}${major}.${minor}.${fix}`
  await _run_shell_command(`git tag ${release_tag}`)
  _print_verbose(`New release is: ${release_tag}`)
  await _update_changelog()
}

async function minor_release() {
  let {description, major, minor, fix} = await current_release()
  minor += 1
  fix = 0
  let release_tag = `${description}${major}.${minor}.${fix}`
  await _run_shell_command(`git tag ${release_tag}`)
  _print_verbose(`New release is: ${release_tag}`)
  await _update_changelog()
}

async function fix_release() {
  let {description, major, minor, fix} = await current_release()
  fix += 1
  let release_tag = `${description}${major}.${minor}.${fix}`
  await _run_shell_command(`git tag ${release_tag}`)
  _print_verbose(`New release is: ${release_tag}`)
  await _update_changelog()
}

async function _update_changelog() {
  // const pwd = process.cwd()
  const project_root = _get_project_root()
  const functions_dir = `${project_root}${path.sep}functions`
  process.chdir(functions_dir)
  await _run_shell_command(`npm run changelog`)
}

function _get_project_root() {
  const cwd = process.cwd()
  return cwd.slice(0, args.PROJECT_NAME.length + cwd.indexOf(args.PROJECT_NAME))
}

//
// ''
// ' Finally, execute the desired command '
// ''
// if args.action is
// not
// None:
//   locals()[args.action]()
// else:
// parser.print_help()
//

if (args.action != null) {
  console.log(args.action)
  const f = module.exports[args.action]
  f()
  // module.exports[args.action]()
  // Reflect.apply(args.action, null, null)
} else {
  parser.print_help()
}


src/parsing.js
/**
 * This module defines functions for parsing command-line arguments.
 *
 * Created on: 07/09/23.
 * @module src/parsing
 * @project js-changelog
 * @author cionzo <cionzo@filotrack.com>
 */

"use strict";

const {ArgumentParser} = require("argparse");
const constants = require("./constants");

/**
 * Get a configured ArgumentParser instance.
 * @returns {ArgumentParser} An ArgumentParser instance configured with command-line argument definitions.
 */
function getParser() {
    const parser = new ArgumentParser({
        description: "Manages releases for this project.",
        add_help: true,
    });

    // Define actions for command-line arguments
    parser.add_argument("-c", "--current", {
        dest: "action",
        action: "store_const",
        const: {name: "currentRelease"},
        help: "returns the current release tag",
    });

    parser.add_argument("-M", "--major", {
        dest: "action",
        action: "store_const",
        const: {name: "createRelease", release: constants.RELEASE_TYPES.MAJOR},
        help: "creates a new major release",
    });
    parser.add_argument("-m", "--minor", {
        dest: "action",
        action: "store_const",
        const: {name: "createRelease", release: constants.RELEASE_TYPES.MINOR},
        help: "creates a new minor release",
    });
    parser.add_argument("-p", "--patch", {
        dest: "action",
        action: "store_const",
        const: {name: "createRelease", release: constants.RELEASE_TYPES.PATCH},
        help: "creates a new patch release",
    });


    parser.add_argument("-cl", "--changelog", {
        dest: "action",
        action: "store_const",
        const: {name: "updateChangelog"},
        help: "updates the changelog file",
    });

    // Define additional arguments
    parser.add_argument("-clo", "--changelog-output", {
        dest: "changelog_output",
        action: "store",
        default: constants.CHANGELOG_PATH,
        help: `the path to the changelog file (defaults to ${constants.CHANGELOG_PATH})`,
    });
    parser.add_argument("-clt", "--changelog-template", {
        dest: "changelog_template",
        action: "store",
        default: constants.CHANGELOG_TEMPLATE_PATH,
        help: "the path to a custom changelog template file",
    });
    parser.add_argument("-pj", "--package-json", {
        dest: "package_json",
        action: "store",
        default: constants.PACKAGE_JSON_PATH,
        help: `the path to the package.json file (defaults to ${constants.PACKAGE_JSON_PATH})`,
    });

    return parser;
}

/**
 * Process command-line arguments and modify them if needed.
 * @param {object} args - Parsed command-line arguments.
 * @returns {void} Nothing.
 */
function processArgs(args) {
    if (args.action && args.action.name  === "currentRelease") {
        args.verbose = true; // Set verbose to true when --current is passed
    }
}

module.exports = {
    getParser,
    processArgs,
};
